MIR: Don't generate 3-armed boolean switch from match.

luqmana opened the pull request:
Fixes #33540.

Snippet from issue:
```Rust
fn foo(x: bool, y: bool) -> u32 {
    match (x, y) {
         (false, _) => 0,
         (_, false) => 1,
         (true, true) => 2,
    }
}
```

Generated MIR:
```
fn foo(arg0: bool, arg1: bool) -> u32 {
    let var0: bool;                      // "x" in scope 1 at 3bbm.rs:17:8: 17:9
    let var1: bool;                      // "y" in scope 1 at 3bbm.rs:17:17: 17:18
    let mut tmp0: (bool, bool);
    let mut tmp1: bool;
    let mut tmp2: bool;
    let mut tmp3: (&'static str, &'static str, u32);
    let mut tmp4: &'static (&'static str, &'static str, u32);

    bb0: {
        var0 = arg0;                     // scope 1 at 3bbm.rs:17:8: 17:9
        var1 = arg1;                     // scope 1 at 3bbm.rs:17:17: 17:18
        tmp1 = var0;                     // scope 5 at 3bbm.rs:18:12: 18:13
        tmp2 = var1;                     // scope 6 at 3bbm.rs:18:15: 18:16
        tmp0 = (tmp1, tmp2);             // scope 4 at 3bbm.rs:18:11: 18:17
        if((tmp0.0: bool)) -> [true: bb4, false: bb1]; // scope 3 at 3bbm.rs:19:10: 19:15
    }

    bb1: {
        return = const 0u32;             // scope 10 at 3bbm.rs:19:23: 19:24
        goto -> bb7;                     // scope 3 at 3bbm.rs:18:5: 22:6
    }

    bb2: {
        return = const 1u32;             // scope 11 at 3bbm.rs:20:23: 20:24
        goto -> bb7;                     // scope 3 at 3bbm.rs:18:5: 22:6
    }

    bb3: {
        return = const 2u32;             // scope 12 at 3bbm.rs:21:25: 21:26
        goto -> bb7;                     // scope 3 at 3bbm.rs:18:5: 22:6
    }

    bb4: {
        if((tmp0.1: bool)) -> [true: bb5, false: bb2]; // scope 3 at 3bbm.rs:20:13: 20:18
    }

    bb5: {
        if((tmp0.0: bool)) -> [true: bb3, false: bb6]; // scope 3 at 3bbm.rs:21:10: 21:14
    }

    bb6: {
        tmp4 = promoted0;                // scope 3 at 3bbm.rs:18:5: 22:6
        core::panicking::panic(tmp4);    // scope 3 at 3bbm.rs:18:5: 22:6
    }

    bb7: {
        return;                          // scope 0 at 3bbm.rs:17:1: 23:2
    }
}
```

Not sure about this approach. I was also thinking maybe just a standalone pass?

cc @arielb1, @nagisa

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

nikomatsakis commented on src/librustc_mir/build/matches/test.rs:
```diff
@@ -162,12 +162,19 @@ impl<'a,'tcx> Builder<'a,'tcx> {
-                           .chain(Some(otherwise))
-                           .collect();
+                let targets: Vec<_> = match switch_ty.sty {
+                    ty::TyBool if options.len() == 2
```
seems...ok to me :) might be nice to be a touch more general (e.g., if there is an exhaustive `u8`)

nikomatsakis replied:
well, this commit alone is not enough, in that it invalidates the `SwitchInt` terminator's invariants. e.g. this comment on `targets` says:

```rust
         /// Possible branch sites. The length of this vector should be
        /// equal to the length of the `values` vector plus 1 -- the
        /// extra item is the block to branch to if none of the values
        /// fit.
        targets: Vec<BasicBlock>,
```

but here you are not generating the otherwise block. It'd be better to change terminator in that case I imagine?

nagisa commented:
This is already something which would break https://github.com/rust-lang/rust/pull/33566.

If you want to land this you’ll have to fix https://github.com/rust-lang/rust/issues/33567 in tandem.

luqmana commented:
Ok, changed SwitchInt to explicitly separate out the otherwise block and put it behind an `Option`. Also, made sure it actually generates valid LLVM IR this time and added a test :P

arielb1 commented:
I *think* we want to generate an `If` for booleans.

nikomatsakis commented:
@arielb1 

>  I think we want to generate an If for booleans.

Or else change `If` to `SwitchInt`?

nagisa commented:
> Or else change If to SwitchInt?

While not necessarily incorrect, that would result in a weird branch in trans since LLVM wants a `br` in boolean case.

luqmana commented:
We could definitely just take `TestKind::SwitchInt` to `TerminatorKind::If` for `bool` types

nikomatsakis commented:
@luqmana 

> We could definitely just take TestKind::SwitchInt to TerminatorKind::If for bool types

I'd be happier with that...

nikomatsakis commented:
...though I imagine we could make `SwitchInt` examine the type of the "int" it is matching against?

luqmana commented:
@nikomatsakis 

> ...though I imagine we could make `SwitchInt` examine the type of the "int" it is matching against?

As in, SwitchInt during `build::matches::test`?

nikomatsakis commented:
@luqmana I think I meant in the trans code. But I'm happy either with using a distinct `If` terminator or with melding `SwitchInt` and `If`.

nikomatsakis commented:
@luqmana do you plan to update this PR to one of those two strategies? :)

bors commented:
:umbrella: The latest upstream changes (presumably #33667) made this pull request unmergeable. Please resolve the merge conflicts.

luqmana commented:
@nikomatsakis Yup sorry, just got occupied with some other work. Anyways, updated the patch to just generate an If terminator for boolean types.

nikomatsakis commented on src/test/run-pass/exhaustive-bool-match-sanity.rs:
```diff
@@ -0,0 +1,24 @@
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+fn foo(x: bool, y: bool) -> u32 {
```
I guess this test should have `#[rustc_mir]`? And a comment explaining what it's all about?

luqmana commented:
Added attribute and comment to test.

arielb1 commented:
@bors r+

bors commented:
:pushpin: Commit a97f6b3 has been approved by `arielb1`

<!-- @bors r=arielb1 a97f6b35ac0532866763c3fa604460497fd3b635 -->

bors commented:
:hourglass: Testing commit a97f6b3 with merge 12d1653...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/9389), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/9427), [auto-linux-32cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-32cross-opt/builds/507), [auto-linux-64-cargotest](http://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/695), [auto-linux-64-cross-armhf](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-armhf/builds/493), [auto-linux-64-cross-armsf](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-armsf/builds/499), [auto-linux-64-cross-freebsd](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/498), [auto-linux-64-cross-netbsd](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-netbsd/builds/503), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/2655), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/9365), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/9348), [auto-linux-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-mir/builds/938), [auto-linux-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/1321), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/9344), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/2539), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/4499), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/9405), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/9409), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/9374), [auto-mac-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/1328), [auto-mac-cross-ios-opt](http://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/515), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/4446), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/4453), [auto-win-gnu-32-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/1339), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/4447), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/4434), [auto-win-msvc-32-cross-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-cross-opt/builds/800), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/3652), [auto-win-msvc-64-cargotest](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/700), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/4446), [auto-win-msvc-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-mir/builds/944), [auto-win-msvc-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/1340)
